release/Earthfile

```diff
@@ -3,6 +3,57 @@ FROM alpine:3.13
 RUN apk add --update --no-cache \
     curl
 
+all:
+    LOCALLY
+    ARG RELEASE_TAG
+    ARG GITHUB_USER=earthly
+    ARG DOCKERHUB_USER=earthly
+    ARG EARTHLY_REPO=earthly
+    ARG BREW_REPO=homebrew-earthly
+    RUN test -n "$RELEASE_TAG" || (echo "ERROR: RELEASE_TAG is not set" && exit 1)
+
+    # validate args: if doing a prod-release make sure no args are changed
+    # otherwise if a test release is detected, make sure all users are changed
+    # (this is to prevent a case where only the github user is changed, but dockerhub is not, and the dockerhub binary is accidentally overwritten)
+    RUN set -e; \
+        test -n "$RELEASE_TAG" || (echo "ERROR: RELEASE_TAG is not set" && exit 1); \
+        (echo "$RELEASE_TAG" | grep '^v[0-9]\+.[0-9]\+.[0-9]\+$' > /dev/null) || (echo "ERROR: RELEASE_TAG must be formatted as v1.2.3; instead got \"$RELEASE_TAG\""; exit 1); \
+        if [ "$GITHUB_USER" = "earthly" ] && [ "$DOCKERHUB_USER" = "earthly" ]; then \
+          test "$BREW_REPO" = "homebrew-earthly" || (echo "ERROR: prod release detected; expecting BREW_REPO=earthly, but got $BREW_REPO" && exit 1); \
+          test "$EARTHLY_REPO" = "earthly" || (echo "ERROR: prod release detected; expecting EARTHLY_REPO=earthly, but got $EARTHLY_REPO" && exit 1); \
+          test "$EARTHLY_GIT_BRANCH" = "main" || (echo "ERROR: performing release on non-main branch" && exit 1); \
+        else \
+          test "$GITHUB_USER" != "earthly" || (echo "ERROR: non-prod release detected but got GITHUB_USER=$GITHUB_USER" && exit 1); \
+          test "$DOCKERHUB_USER" != "earthly" || (echo "ERROR: non-prod release detected but got DOCKERHUB_USER=$DOCKERHUB_USER" && exit 1); \
+        fi; \
+        echo "performing release to repo=$GITHUB_USER/$EARTHLY_REPO; homebrew=$GITHUB_USER/$BREW_REPO; dockerhub=$DOCKERHUB_USER"
+
+    # terrible unsafe hack to determine --push flag
+    IF ps auxw | grep -v grep | grep earthly | grep -e --push
+        ARG PUSH="--push"
+    ELSE
+        ARG PUSH=""
+    END
```

Comment:
This doesn't look great :D

Should we add a builtin arg `EARTHLY_PUSH` which would be set to either `true` or `false` accordingly?

Reply:
it's downright terrible :grimacing: 

That's a very interesting idea regarding that flag. I could see us abusing it like:

```
IF $EARTHLY_PUSH
    RUN --no-cache do-a-push
END
RUN echo a non-push command comes after.
```

That alone might allow us to get rid of the `../earthly` calling `../earthly` hack too.

Would allowing us (and our users) to adopt such a pattern cause any issues?

I don't think so, as a user could currently introduce their own `--build-arg MY_PUSH_FLAG=true` to implement something very similar.

Reply:
this LOCALLY target has been moved into a seperate `./release.sh` script which will call `earthly` several times. We will tackle supporting this via a single earthly command in the future.